command,runner: limit environ size when starting command

adambabik opened the pull request:
This change ensures that the size of individual environment variables, as well as all environment variables (environ) passed to a command, are within limits.

Env can come from various sources like system, session, and project. The  problem with exceeding the limit is typically caused by storing the last output of a previous command in a special env called `__`. The algorithm recognized this and trims the prefix of `__`. While executing the command, the last output is stored in a ring buffer anyway so the behaviour stays the same. If the variable is not present or trimming is not enough, the error is logged but still the command is tried to be executed.

This change also fixes a problem of trimming large outputs when executing commands via the `runnerv2` service.

Relates to https://github.com/stateful/runme/issues/648

sourishkrout commented:
Something's not quite right:

https://gist.github.com/sourishkrout/3c8153387a81a30ec5e0d55e0d601ba1

On the notebook (client-side), I ran this where downloaded the 1m records users file from https://jsoneditoronline.org/indepth/datasets/json-file-example/

```
du -h /Users/sourishkrout/Downloads/users_1m.json

# Ran on 2024-08-22 15:27:25-07:00 for 178ms exited with 0
296M/Users/sourishkrout/Downloads/users_1m.json
```

```
cat /Users/sourishkrout/Downloads/users_1m.json

# Ran on 2024-08-22 15:27:27-07:00 for 2.165s exited with 0
[... ultra large output]
```

sourishkrout commented:
Tracing the size against the limit is interesting. Are we comparing the right units on both sides @adambabik?

```sh
go run . server --address 127.0.0.1:9999 --tls /tmp/runme/tls 2>/dev/null

# Ran on 2024-08-22 15:31:45-07:00 for 2.059s
size:  3746 MaxEnvironSizeInBytes:  128000000
size:  3886 MaxEnvironSizeInBytes:  128000000
size:  3886 MaxEnvironSizeInBytes:  128000000
size:  3799 MaxEnvironSizeInBytes:  128000000
size:  3939 MaxEnvironSizeInBytes:  128000000
size:  3939 MaxEnvironSizeInBytes:  128000000
size:  2100901 MaxEnvironSizeInBytes:  128000000
size:  2101041 MaxEnvironSizeInBytes:  128000000
size:  2101041 MaxEnvironSizeInBytes:  128000000
```

sourishkrout reviewed (changes requested):
Long story short is that https://github.com/stateful/runme/issues/648 produces the same error. Please see analysis in the comments.

sourishkrout commented:
> ```
> size:  2101041 MaxEnvironSizeInBytes:  128000000
> ```

Upon further investigation is looks like the entire file never makes it into the env store. It appears to truncate it somehow. The chunks seem to closely track the `msgBufferSize` of its message ring buffer. But that's just eyeballing it. I'm confused 🤷.

adambabik commented:
@sourishkrout just so I understand the test scenario. In one cell, you run `cat large_file.json` and set a flag to store its stdout in the environment variable. In the second cell, you count the bytes of the value of that environment variable. The size limits you collected above are the sizes of the value of the env, correct?

sourishkrout commented:
> @sourishkrout just so I understand the test scenario. In one cell, you run `cat large_file.json` and set a flag to store its stdout in the environment variable. In the second cell, you count the bytes of the value of that environment variable. The size limits you collected above are the sizes of the value of the env, correct?

Correct. The size is reported (on the Go kernel side) when the ENV is populated (i.e. `$__` that's too large) before executing cells. Please see: https://gist.github.com/sourishkrout/66ca32c2e806d1783c19080d6d4cc8cc

However, what I'm noticing `v1` vs `v2` appears that there's an issue with cat'ing the file in the first place. It seems to truncate prematurely (please see video below) in `v2`.


https://github.com/user-attachments/assets/0e8f6eaa-5836-4835-b736-9c08bbd3ccb9

sourishkrout commented:
Btw, it appears the last x characters of `v1` matches the file as opposed to `v2`:

https://gist.github.com/sourishkrout/7002a783292b9184df9c88a9dbbc4f32#file-01j68939kbe1y6k9z0h4f7mn9n-md

Possibly something off with the ring buffer in `runnerv2`?

sourishkrout commented:
Also, for easier "debugging". Here's my log from just cat-ing the large file:

```
2024-08-26T14:28:50.651-0700    INFO    runnerv2service/service_execute.go:34   received initial request        {"id": "01J689EX43MBZGYZKJPBEHYG7R", "req": "config:{program_name:\"/bin/zsh\"  directory:\"/Users/sourishkrout/Projects/stateful/oss/vscode-runme/examples\"  language_id:\"sh\"  env:\"RUNME_ID=01J56QWV12GPV94QWBFYVG2G2W\"  env:\"RUNME_RUNNER=v2\"  env:\"TERM=xterm-256color\"  commands:{items:\"cat /Users/sourishkrout/Downloads/users_1m.json\"  items:\"\"}  interactive:true  mode:COMMAND_MODE_INLINE  known_id:\"01J56QWV12GPV94QWBFYVG2G2W\"  known_name:\"cat-userssourishkroutdownload\"}  winsize:{rows:10  cols:159}  session_id:\"01J689ET9B4DB1J5NNNBXQPAHQ\"  store_stdout_in_env:true"}
2024-08-26T14:28:50.652-0700    DEBUG   InlineShellCommand      command/command_inline_shell.go:36      inline shell script     {"id": "01J689EX43MBZGYZKJPBEHYG7R", "instanceID": "01J689EX6W80S0R2EJK90HSHYR", "script": "set -e -o pipefail\n\n /var/folders/c3/5r0t1nzs7sbfpxjgbc6n3ss40000gn/T/go-build3711462266/b001/exe/v3 env dump --insecure > /var/folders/c3/5r0t1nzs7sbfpxjgbc6n3ss40000gn/T/runme-1701673126/env_pre.fifo\n __cleanup() {\nrv=$?\n/var/folders/c3/5r0t1nzs7sbfpxjgbc6n3ss40000gn/T/go-build3711462266/b001/exe/v3 env dump --insecure > /var/folders/c3/5r0t1nzs7sbfpxjgbc6n3ss40000gn/T/runme-1701673126/env_post.fifo\nexit $rv\n}\n trap -- \"__cleanup\" EXIT\ncat /Users/sourishkrout/Downloads/users_1m.json\n\n"}
2024-08-26T14:28:50.653-0700    INFO    VirtualCommand  command/command_virtual.go:73   detected program path and arguments     {"id": "01J689EX43MBZGYZKJPBEHYG7R", "instanceID": "01J689EX6W80S0R2EJK71133X7", "program": "/bin/zsh", "args": ["-c", "set -e -o pipefail\n\n /var/folders/c3/5r0t1nzs7sbfpxjgbc6n3ss40000gn/T/go-build3711462266/b001/exe/v3 env dump --insecure > /var/folders/c3/5r0t1nzs7sbfpxjgbc6n3ss40000gn/T/runme-1701673126/env_pre.fifo\n __cleanup() {\nrv=$?\n/var/folders/c3/5r0t1nzs7sbfpxjgbc6n3ss40000gn/T/go-build3711462266/b001/exe/v3 env dump --insecure > /var/folders/c3/5r0t1nzs7sbfpxjgbc6n3ss40000gn/T/runme-1701673126/env_post.fifo\nexit $rv\n}\n trap -- \"__cleanup\" EXIT\ncat /Users/sourishkrout/Downloads/users_1m.json\n\n"]}
2024-08-26T14:28:50.653-0700    INFO    VirtualCommand  command/command_virtual.go:93   starting        {"id": "01J689EX43MBZGYZKJPBEHYG7R", "instanceID": "01J689EX6W80S0R2EJK71133X7", "config": "program_name:\"/bin/zsh\"  arguments:\"-c\"  arguments:\"set -e -o pipefail\\n\\n /var/folders/c3/5r0t1nzs7sbfpxjgbc6n3ss40000gn/T/go-build3711462266/b001/exe/v3 env dump --insecure > /var/folders/c3/5r0t1nzs7sbfpxjgbc6n3ss40000gn/T/runme-1701673126/env_pre.fifo\\n __cleanup() {\\nrv=$?\\n/var/folders/c3/5r0t1nzs7sbfpxjgbc6n3ss40000gn/T/go-build3711462266/b001/exe/v3 env dump --insecure > /var/folders/c3/5r0t1nzs7sbfpxjgbc6n3ss40000gn/T/runme-1701673126/env_post.fifo\\nexit $rv\\n}\\n trap -- \\\"__cleanup\\\" EXIT\\ncat /Users/sourishkrout/Downloads/users_1m.json\\n\\n\"  directory:\"/Users/sourishkrout/Projects/stateful/oss/vscode-runme/examples\"  commands:{items:\"cat /Users/sourishkrout/Downloads/users_1m.json\"  items:\"\"}  interactive:true  mode:COMMAND_MODE_INLINE"}
2024-08-26T14:28:50.654-0700    INFO    VirtualCommand  command/command_virtual.go:97   started {"id": "01J689EX43MBZGYZKJPBEHYG7R", "instanceID": "01J689EX6W80S0R2EJK71133X7"}
2024-08-26T14:28:50.655-0700    INFO    VirtualCommand  command/command_virtual.go:163  waiting for finish      {"id": "01J689EX43MBZGYZKJPBEHYG7R", "instanceID": "01J689EX6W80S0R2EJK71133X7"}
2024-08-26T14:28:50.695-0700    INFO    runnerv2service/execution.go:130        detected MIME type      {"id": "01J689EX43MBZGYZKJPBEHYG7R", "mime": "application/json"}
2024-08-26T14:28:52.567-0700    INFO    VirtualCommand  command/command_virtual.go:165  finished        {"id": "01J689EX43MBZGYZKJPBEHYG7R", "instanceID": "01J689EX6W80S0R2EJK71133X7"}
2024-08-26T14:28:52.567-0700    INFO    VirtualCommand  command/command_virtual.go:168  closed IO       {"id": "01J689EX43MBZGYZKJPBEHYG7R", "instanceID": "01J689EX6W80S0R2EJK71133X7"}
2024-08-26T14:28:52.567-0700    INFO    VirtualCommand  command/command_virtual.go:133  copied from pty to stdout       {"id": "01J689EX43MBZGYZKJPBEHYG7R", "instanceID": "01J689EX6W80S0R2EJK71133X7", "count": 310365941}
2024-08-26T14:28:52.567-0700    INFO    VirtualCommand  command/command_virtual.go:173  waiting IO goroutines   {"id": "01J689EX43MBZGYZKJPBEHYG7R", "instanceID": "01J689EX6W80S0R2EJK71133X7"}
2024-08-26T14:28:52.567-0700    INFO    VirtualCommand  command/command_virtual.go:107  copied from stdin to pty        {"id": "01J689EX43MBZGYZKJPBEHYG7R", "instanceID": "01J689EX6W80S0R2EJK71133X7", "count": 0}
2024-08-26T14:28:52.567-0700    INFO    VirtualCommand  command/command_virtual.go:175  finished waiting for IO goroutines      {"id": "01J689EX43MBZGYZKJPBEHYG7R", "instanceID": "01J689EX6W80S0R2EJK71133X7"}
2024-08-26T14:28:52.567-0700    INFO    InlineShellCommand      command/command_inline_shell.go:52      collecting the environment after the script execution   {"id": "01J689EX43MBZGYZKJPBEHYG7R", "instanceID": "01J689EX6W80S0R2EJK90HSHYR"}
2024-08-26T14:28:52.568-0700    INFO    InlineShellCommand      command/command_inline_shell.go:54      collected the environment after the script execution    {"id": "01J689EX43MBZGYZKJPBEHYG7R", "instanceID": "01J689EX6W80S0R2EJK90HSHYR"}
2024-08-26T14:28:52.568-0700    INFO    runnerv2service/execution.go:161        command finished        {"id": "01J689EX43MBZGYZKJPBEHYG7R", "exitCode": 0}
2024-08-26T14:28:52.568-0700    INFO    runnerv2service/execution.go:252        closed stdin writer     {"id": "01J689EX43MBZGYZKJPBEHYG7R"}
2024-08-26T14:28:52.568-0700    INFO    runnerv2service/execution.go:255        closed stdout writer    {"id": "01J689EX43MBZGYZKJPBEHYG7R"}
2024-08-26T14:28:52.568-0700    INFO    runnerv2service/execution.go:258        closed stderr writer    {"id": "01J689EX43MBZGYZKJPBEHYG7R"}
2024-08-26T14:28:52.569-0700    INFO    runnerv2service/service_execute.go:125  command finished        {"id": "01J689EX43MBZGYZKJPBEHYG7R", "exitCode": 0}
2024-08-26T14:28:52.569-0700    INFO    runnerv2service/service_execute.go:100  received request        {"id": "01J689EX43MBZGYZKJPBEHYG7R", "req": "<nil>", "error": "rpc error: code = Canceled desc = context canceled"}
2024-08-26T14:28:52.569-0700    INFO    runnerv2service/service_execute.go:112  stream canceled after the process finished; ignoring    {"id": "01J689EX43MBZGYZKJPBEHYG7R"}
```

sourishkrout commented:
Following up on the above. This is very interesting @adambabik. When I reported the length of stdout in `v1` vs `v2`:

with `v1`:
```
[...]
2024-08-26T14:33:50.420-0700    DEBUG   runner/service.go:469   sending data    {"_id": "01J689QYQFNPZ89CDDFTKHCD30", "knownID": "01J56QWV12GPV94QWBFYVG2G2W", "knownName": "cat-userssourishkroutdownload", "lenStdout": 32000, "lenStderr": 0}
2024-08-26T14:33:50.426-0700    DEBUG   runner/service.go:469   sending data    {"_id": "01J689QYQFNPZ89CDDFTKHCD30", "knownID": "01J56QWV12GPV94QWBFYVG2G2W", "knownName": "cat-userssourishkroutdownload", "lenStdout": 32000, "lenStderr": 0}
2024-08-26T14:33:50.428-0700    DEBUG   runner/service.go:469   sending data    {"_id": "01J689QYQFNPZ89CDDFTKHCD30", "knownID": "01J56QWV12GPV94QWBFYVG2G2W", "knownName": "cat-userssourishkroutdownload", "lenStdout": 32000, "lenStderr": 0}
2024-08-26T14:33:50.429-0700    DEBUG   runner/service.go:469   sending data    {"_id": "01J689QYQFNPZ89CDDFTKHCD30", "knownID": "01J56QWV12GPV94QWBFYVG2G2W", "knownName": "cat-userssourishkroutdownload", "lenStdout": 32000, "lenStderr": 0}
2024-08-26T14:33:50.433-0700    DEBUG   runner/service.go:469   sending data    {"_id": "01J689QYQFNPZ89CDDFTKHCD30", "knownID": "01J56QWV12GPV94QWBFYVG2G2W", "knownName": "cat-userssourishkroutdownload", "lenStdout": 32000, "lenStderr": 0}
2024-08-26T14:33:50.439-0700    DEBUG   runner/service.go:469   sending data    {"_id": "01J689QYQFNPZ89CDDFTKHCD30", "knownID": "01J56QWV12GPV94QWBFYVG2G2W", "knownName": "cat-userssourishkroutdownload", "lenStdout": 32000, "lenStderr": 0}
2024-08-26T14:33:50.440-0700    DEBUG   runner/service.go:469   sending data    {"_id": "01J689QYQFNPZ89CDDFTKHCD30", "knownID": "01J56QWV12GPV94QWBFYVG2G2W", "knownName": "cat-userssourishkroutdownload", "lenStdout": 32000, "lenStderr": 0}
2024-08-26T14:33:50.445-0700    DEBUG   runner/service.go:469   sending data    {"_id": "01J689QYQFNPZ89CDDFTKHCD30", "knownID": "01J56QWV12GPV94QWBFYVG2G2W", "knownName": "cat-userssourishkroutdownload", "lenStdout": 32000, "lenStderr": 0}
2024-08-26T14:33:50.445-0700    DEBUG   runner/service.go:469   sending data    {"_id": "01J689QYQFNPZ89CDDFTKHCD30", "knownID": "01J56QWV12GPV94QWBFYVG2G2W", "knownName": "cat-userssourishkroutdownload", "lenStdout": 24053, "lenStderr": 0}
2024-08-26T14:33:50.447-0700    ERROR   runner/service.go:531   could not set environment variable, environment size limit exceeded     {"_id": "01J689QYQFNPZ89CDDFTKHCD30", "knownID": "01J56QWV12GPV94QWBFYVG2G2W", "knownName": "cat-userssourishkroutdownload"}
2024-08-26T14:33:50.447-0700    INFO    runner/service.go:546   sending the final response with exit code       {"_id": "01J689QYQFNPZ89CDDFTKHCD30", "knownID": "01J56QWV12GPV94QWBFYVG2G2W", "knownName": "cat-userssourishkroutdownload", "exitCode": 0}
2024-08-26T14:33:50.449-0700    INFO    runner/service.go:399   stream canceled after the process finished; ignoring    {"_id": "01J689QYQFNPZ89CDDFTKHCD30", "knownID": "01J56QWV12GPV94QWBFYVG2G2W", "knownName": "cat-userssourishkroutdownload"}
[...]
```

with `v2`:
```
[...]
2024-08-26T14:35:31.401-0700    INFO    VirtualCommand  command/command_virtual.go:97   started {"id": "01J689V4C2ENQ45WDQ0X2A9591", "instanceID": "01J689V4J7706NCEPQZYC04NX8"}
2024-08-26T14:35:31.411-0700    INFO    VirtualCommand  command/command_virtual.go:163  waiting for finish      {"id": "01J689V4C2ENQ45WDQ0X2A9591", "instanceID": "01J689V4J7706NCEPQZYC04NX8"}
2024-08-26T14:35:31.425-0700    INFO    runnerv2service/execution.go:130        detected MIME type      {"id": "01J689V4C2ENQ45WDQ0X2A9591", "mime": "application/json"}
2024-08-26T14:35:31.425-0700    DEBUG   runnerv2service/execution.go:137        sending stdout data     {"id": "01J689V4C2ENQ45WDQ0X2A9591", "lenStdout": 33792}
2024-08-26T14:35:31.425-0700    DEBUG   runnerv2service/execution.go:137        sending stdout data     {"id": "01J689V4C2ENQ45WDQ0X2A9591", "lenStdout": 34816}
2024-08-26T14:35:31.425-0700    DEBUG   runnerv2service/execution.go:137        sending stdout data     {"id": "01J689V4C2ENQ45WDQ0X2A9591", "lenStdout": 11264}
2024-08-26T14:35:31.426-0700    DEBUG   runnerv2service/execution.go:137        sending stdout data     {"id": "01J689V4C2ENQ45WDQ0X2A9591", "lenStdout": 12288}
2024-08-26T14:35:31.426-0700    DEBUG   runnerv2service/execution.go:137        sending stdout data     {"id": "01J689V4C2ENQ45WDQ0X2A9591", "lenStdout": 38912}
2024-08-26T14:35:31.427-0700    DEBUG   runnerv2service/execution.go:137        sending stdout data     {"id": "01J689V4C2ENQ45WDQ0X2A9591", "lenStdout": 35840}
2024-08-26T14:35:31.427-0700    DEBUG   runnerv2service/execution.go:137        sending stdout data     {"id": "01J689V4C2ENQ45WDQ0X2A9591", "lenStdout": 31744}
2024-08-26T14:35:31.437-0700    DEBUG   runnerv2service/execution.go:137        sending stdout data     {"id": "01J689V4C2ENQ45WDQ0X2A9591", "lenStdout": 615424}
2024-08-26T14:35:31.441-0700    DEBUG   runnerv2service/execution.go:137        sending stdout data     {"id": "01J689V4C2ENQ45WDQ0X2A9591", "lenStdout": 299008}
2024-08-26T14:35:31.454-0700    DEBUG   runnerv2service/execution.go:137        sending stdout data     {"id": "01J689V4C2ENQ45WDQ0X2A9591", "lenStdout": 557056}
2024-08-26T14:35:31.478-0700    DEBUG   runnerv2service/execution.go:137        sending stdout data     {"id": "01J689V4C2ENQ45WDQ0X2A9591", "lenStdout": 427008}
2024-08-26T14:35:33.307-0700    INFO    VirtualCommand  command/command_virtual.go:165  finished        {"id": "01J689V4C2ENQ45WDQ0X2A9591", "instanceID": "01J689V4J7706NCEPQZYC04NX8"}
[...]
```

adambabik commented:
@sourishkrout thanks for the details. I found the bug and fixed it. There were two problems:
1. Using `io.LimitedReader` incorrectly which resulted in a shorter output. I remember that it was done for safety to limit the maximum output. I removed it for now.
2. Ring buffer was actually used as stdout and stderr in the `execution` struct in the service and it also could cause missing some output, provided that the command generated it quickly enough (`cat` was enough to reproduce). I replaced it with a regular buffer.

Please test it again. I added unit tests that replicated it and now they are passing.

sourishkrout commented:
Excellent. Thanks for the fix, @adambabik! The output now appears correctly in the output cell (integration with extension) 👍 . However, the original issue that subsequent cells are not running (screenshot below) due to the ENV being too large persists.

I suggest you merge this fix and address the remaining issue in a separate PR. Without well-known limits for every OS's ENV size, we should pick a hard limit for storing the last output and trim it accordingly. Trying to be too clever here will likely make it hard to reason about scenarios where the limit is exceeded and/or not even being aware of problems unless reported by users. Wdyt @adambabik?

sourishkrout reviewed (approved):
Let's merge an use separate PR for remaining issue as per my comment.

adambabik commented:
@sourishkrout thanks for the review. You're right, I was focusing on the second problem, and the environment limit was not fixed. I agree that being "clever" here is useless and a reliable solution is much more complex and really requires detecting the boundary case by case. It's not a simple function of operating system. I have found and run this on my machine:
```
getconf -a | grep ARG_MAX
ARG_MAX: 1048576
```

However, `ARG_MAX` is a guaranteed minimum. I guess we can proceed with 2 MB which should be a quite safe value.

adambabik commented:
@sourishkrout finally, I found limits that worked for macOS Ventura and Debian. I think we can merge it now. If you want to further tweak the numbers, take a look [here](https://github.com/stateful/runme/pull/654#discussion_r1750719511).

sourishkrout commented on internal/command/env_store.go:
```diff
@@ -4,13 +4,16 @@ import (
-// 	// +2 for the '=' and '\0' separators
-// 	return len(k) + len(v) + 2
-// }
+const (
```
It works on Sonoma, too 👍. Let's roll with these for now.
